tracer/src/Datadog.Trace/Configuration/ConfigurationSources/StringConfigurationSource.cs

```diff
@@ -75,36 +75,44 @@ public abstract class StringConfigurationSource : IConfigurationSource, ITelemet
+                // we need Trim() before looking for ':' so we can skip entries with no key
                 // (that is, entries with a leading ':', like "<empty or whitespace>:value")
-                var trimmedEntry = entry.TrimStart();
+                var trimmedEntry = entry.Trim();
+                if (trimmedEntry.Length == 0 || trimmedEntry.StartsWith(":"))
+                {
+                    continue;
+                }
+                else if (enableHeaderTagsBehaviors && trimmedEntry.EndsWith(":"))
+                {
+                    // When parsing header tags, any input trailing colons is invalid
+                    continue;
+                }
 
-                if (trimmedEntry.Length > 0)
+                var colonIndex = enableHeaderTagsBehaviors switch
+                {
+                    false => trimmedEntry.IndexOf(':'), // In the general case, we split on the first colon
+                    true => trimmedEntry.LastIndexOf(':'), // However, for header tags parsing is recommended to split on last colon
+                };
+
+                if (colonIndex < 0 && enableHeaderTagsBehaviors)
+                {
+                    // entries with no colon are allowed (e.g. "key1, key2:value2, key3"),
+                    // it's a key with no value.
+                    var key = trimmedEntry;
+                    dictionary[key] = string.Empty;
+                }
+                else if (colonIndex > 0)
                 {
-                    // colonIndex == 0 is a leading colon, not valid
-                    var colonIndex = trimmedEntry.IndexOf(':');
-
-                    if (colonIndex < 0 && allowOptionalMappings)
-                    {
-                        // entries with no colon are allowed (e.g. "key1, key2:value2, key3"),
-                        // it's a key with no value.
-                        // note we already did TrimStart(), so we only need TrimEnd().
-                        var key = trimmedEntry.TrimEnd();
-                        dictionary[key] = string.Empty;
-                    }
-                    else if (colonIndex > 0)
-                    {
-                        // split at the first colon only. any other colons are part of the value.
-                        // if a colon is present with no value, we take the value to be empty (e.g. "key1:, key2: ").
-                        // note we already did TrimStart() on the key, so it only needs TrimEnd().
-                        var key = trimmedEntry.Substring(0, colonIndex).TrimEnd();
-                        var value = trimmedEntry.Substring(colonIndex + 1).Trim();
-                        dictionary[key] = value;
-                    }
+                    // if a colon is present with no value, we take the value to be empty (e.g. "key1:, key2: ").
+                    // note we already did TrimStart() on the key, so it only needs TrimEnd().
```

Comment:
nitpick, but this comment is a tad misleading now

We did a `Trim` on the whole entry (which is essentially a `TrimStart` on the key), but when we split it we'd need to do a `TrimEnd`

```suggestion
                    // note we already did Trim() on the entire entry, so the key portion only needs TrimEnd().
```